src/event-db/src/types/snapshot.rs

```diff
@@ -0,0 +1,155 @@
+            type Value = SnapshotVersion;
+            fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result {
+                formatter.write_str(
+                    r#"Expect one of the following options: "latest", "25", "Fund 10" etc."#,
+                )
+            }
+            fn visit_borrowed_str<E>(self, v: &'de str) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                if v == "latest" {
+                    Ok(Self::Value::Latest)
+                } else {
+                    match v.parse::<i32>() {
+                        Ok(number) => Ok(Self::Value::Number(number)),
+                        _ => Ok(Self::Value::Name(v.to_string())),
+                    }
+                }
+            }
+        }
+        deserializer.deserialize_any(SnapshotVersionVisitor)
+    }
+}
+
+#[derive(Serialize, Clone, Default)]
+pub struct VoterInfo {
+    pub voting_power: i64,
+    pub voting_group: String,
+    pub delegations_power: i64,
+    pub delegations_count: i64,
+    pub voting_power_saturation: f64,
+}
+
+#[derive(Serialize, Clone)]
+pub struct Voter {
+    pub voter_info: VoterInfo,
+    #[serde(serialize_with = "serialize_systemtime_as_rfc3339")]
+    pub as_at: SystemTime,
+    #[serde(serialize_with = "serialize_systemtime_as_rfc3339")]
+    pub last_updated: SystemTime,
+    pub r#final: bool,
+}
+
+#[derive(Serialize, Clone, Default)]
+pub struct Delegation {
+    pub voting_key: String,
+    pub group: String,
+    pub weight: i32,
+    pub value: i64,
+}
+
+#[derive(Serialize, Clone)]
+pub struct Delegator {
+    pub delegations: Vec<Delegation>,
+    pub raw_power: i64,
+    pub total_power: i64,
+    #[serde(serialize_with = "serialize_systemtime_as_rfc3339")]
+    pub as_at: SystemTime,
+    #[serde(serialize_with = "serialize_systemtime_as_rfc3339")]
+    pub last_updated: SystemTime,
+    pub r#final: bool,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn snapshot_version_json_test() {
+        let snapshot_versions = vec![
+            SnapshotVersion::Latest,
+            SnapshotVersion::Number(10),
+            SnapshotVersion::Name("Fund 10".to_string()),
+        ];
+        let json = serde_json::to_string(&snapshot_versions).unwrap();
+        assert_eq!(json, r#"["latest","10","Fund 10"]"#);
+
+        let decoded: Vec<SnapshotVersion> = serde_json::from_str(&json).unwrap();
+        assert_eq!(decoded, snapshot_versions);
+    }
+
+    #[test]
+    fn voter_json_test() {
+        let voter = Voter {
+            voter_info: VoterInfo {
+                voting_power: 100,
+                voting_group: "rep".to_string(),
+                delegations_power: 100,
+                delegations_count: 1,
+                voting_power_saturation: 0.4,
+            },
+            as_at: SystemTime::UNIX_EPOCH,
+            last_updated: SystemTime::UNIX_EPOCH,
```

Comment:
Does this mean "not updated"? If so, maybe an `Option<SystemTime>` (or `Option<DateTime<Utc>>`) would be clearer?

Reply:
No no no, it is not possible to make it optional, because it is a correspondance field to this one from the db schema https://github.com/input-output-hk/catalyst-core/blob/d2aabde031464526ab2fbaf09f1ef4a60ca81776/src/event-db/migrations/V6__snapshot_tables.sql#L9. Here in tests I just put some test values to test serde

Reply:
Ahh makes sense 👍